Stripping base path from CLI paths

cameel opened the pull request:
Fixes #4702.
Implements the `solc` part of #11408. `solc-js` changes will be submitted in that repo.
~Depends on #11544.~ Merged.

This still work in progress. I'd say It's about 90% done. There's not much implementation left (it's very small overall) but some corner cases are not yet handled or not handled properly:
- [x] Paths with a different case on case-insensitive filesystems.
- [x] Paths starting with `//` (UNC paths).
- [x] Paths using `..` to go beyond filesystem root (boost's normalization does not squash `/../` into `/`).
- [x] Paths with symlinks.
- [x] Removing root name so that paths always start with `/` (rather than drive letter on Windows) when they're on the same partition as CWD.
- [x] Different behavior on some older Boost versions.

cameel committed:
CLI tests for FileReader

cameel committed:
FileReader: Normalize base path and strip it from normalized source paths

chriseth commented on libsolidity/interface/FileReader.cpp:
```diff
@@ -31,9 +31,23 @@ using std::string;
 void FileReader::setSource(boost::filesystem::path const& _path, SourceCode _source)
 {
-	m_sourceCodes[_path.generic_string()] = std::move(_source);
+	boost::filesystem::path normalizedPath = normalizeCLIPathForVFS(_path).generic_string();
```
Why do you call `generic_string` and then re-assign to a `path`?

cameel replied:
Good catch. That's a mistake. It should not be called here, only below.

chriseth commented on libsolidity/interface/FileReader.cpp:
```diff
@@ -92,5 +106,42 @@ ReadCallback::Result FileReader::readFile(string const& _kind, string const& _so
+
+boost::filesystem::path FileReader::stripPathPrefix(boost::filesystem::path _prefix, boost::filesystem::path const& _path)
+{
+	solAssert(isPathPrefix(_prefix, _path), "");
```
If you call `isPathPrefix` anyway, why not do `stripPathIfPrefix`?

chriseth replied:
or `stripPrefixIfPresent`?

cameel replied:
Because I want to reuse it for `--allowed-paths` later. I originally wanted to do that in this PR but then realized that I can't because allowed paths are not normalized currently (#4623). And they need slightly different normalization because we want to resolve symlinks in them while we don't want to do that in base path. Overall, I wanted to focus on that in a separate PR but I already know that I'll need two separate functions so I implemented it like that here.

cameel commented:
Here's a short update on the current status on this PR:
- I dealt with most of the corner cases listed in PR description. The last remaining ones are:
    - Symlinks. On macOS in CI the path to the temporary directory contains a symlink (`/var/folders/` -> `/private/var/folders/`) and Boost sometimes resolves the symlink when the path is relative and starts with `../`. I suspect it might be a bug in its boost version but I need to investigate. This will be tedious because I can reproduce it only in CI.
    - The root path is sometimes `C:/` on Windows where it should be normalized to `/`. This might be a bug on my part.
- I resolved all the CI failures in the refactor PRs it depends on (#11518, #11520, #11544). But as requested by @chriseth I'll put this one directly on `develop` anyway - in a stripped-down version (without CLI tests) - once I get tests to pass.

cameel committed:
createSymlinkIfSupportedByFilesystem(): Add support for directory symlinks used on Windows

cameel commented:
This PR is now working correctly on all platforms. It still depends on the CLI refactor so I have created an equivalent PR without the CLI tests that does not: #11617. These tests turned out to be crucial as I suspected. There were tons of weird cases that the `FileReader` unit tests did not catch (see https://github.com/ethereum/solidity/pull/11617#issuecomment-874309838).

cameel commented:
#11544 has been merged so this is now reviewable.

Note that the only difference between this PR and #11617 are tests in `test/solc/CommandLineInterface.cpp` so if you reviewed that other one already, you only have to take a look at that one extra file.

leonardoalt commented on libsolidity/interface/FileReader.cpp:
```diff
@@ -92,5 +102,140 @@ ReadCallback::Result FileReader::readFile(string const& _kind, string const& _so
+	return normalizedPathNoDotDot;
+}
+
+bool FileReader::isPathPrefix(boost::filesystem::path _prefix, boost::filesystem::path const& _path)
```
`_prefix` is not moved, so could be `const&`

cameel replied:
Like I said in https://github.com/ethereum/solidity/pull/11617#discussion_r670304196, I'm modifying this inside the function which is why it's not `const`.

Well ok, I can change that anyway, I now switched from `remove_filename()` to `parent_path()` which does not modify the path. `const&` added.

leonardoalt commented on libsolidity/interface/FileReader.cpp:
```diff
@@ -92,5 +102,140 @@ ReadCallback::Result FileReader::readFile(string const& _kind, string const& _so
+	return !strippedPath.empty() && *strippedPath.begin() != "..";
+}
+
+boost::filesystem::path FileReader::stripPrefixIfPresent(boost::filesystem::path _prefix, boost::filesystem::path const& _path)
```
`_prefix` is not moved so could be `const&`

cameel replied:
Done: https://github.com/ethereum/solidity/pull/11545#discussion_r682626866

leonardoalt commented on test/solc/Common.cpp:
```diff
@@ -56,3 +56,19 @@ test::OptionsReaderAndMessages test::parseCommandLineAndReadInputFiles(
+
+	options.optimizer.expectedExecutionsPerDeployment = 200;
+	options.modelChecker.initialize = true;
+	options.modelChecker.settings = {
```
Why not `= {}`; ?

cameel replied:
Right, I should have used that. Fixed.

leonardoalt reviewed (commented):
Some small comments, looks good otherwise!

cameel commented:
Old fixups squashed. New changes added in new fixups. Ready for another round.

Please review https://github.com/ethereum/solc-js/pull/533 too, which is the equivalent PR for solc-js. It's much shorter because the normalization there is not as strict.

aarlt commented on docs/path-resolution.rst:
```diff
@@ -313,6 +315,46 @@ interpreted as absolute paths on disk.
+
+- If a path is relative, it is made absolute by prepending the current working directory to it.
+
+  - If the path to the working directory contains symbolic links, they are resolved into actual
```
Is this indention correct?

cameel replied:
Yes. That's a nested bullet point. The parent point speaks about "working directory" and this extra point explains a related detail - it says what happens if the path to the working directory has symbolic links in it. I could make it a separate point but it's only really relevant in the context of relative paths.

And if you mean specifically the RST syntax, the indentation is always long enough to match the bullet point of the parent. So here it's two spaces. If the parent was `1. If a path ...` then you'd need to indent with 3 spaces. With `123. If a path ...` you'd need 5 and so on.

aarlt commented on libsolidity/interface/FileReader.cpp:
```diff
@@ -31,9 +33,17 @@ using std::string;
 
+void FileReader::setBasePath(boost::filesystem::path const& _path)
+{
+	m_basePath = (_path.empty() ? "" : normalizeCLIPathForVFS(_path));
```
What about setting `normalizeCLIPathForVFS(".")` already here, if `_path` is empty? With this `m_basePath` would always contain a valid path.

cameel replied:
The problem is that `""` is a special case that is not equivalent to any specific path.

For example if I use the working dir as you're suggesting, it will break absolute paths - `import "/x/y.sol"` will try to load `x.y.sol` from work dir rather than the absolute `/x/y.sol`.

Using `/` would keep absolute paths working but it would relative paths - `import "x/y.sol"` would try to load `/x.y.sol`.

So these two values are out and I can't think of any other path that would keep `--base-path=""` working as it does now.

cameel commented:
While working on #11409 I found a small bug introduced in this PR: standard input was not being handled correctly in combination with base path. `<stdin>` was converted to an absolute path as if it were a file name. I added a fixup that solves this and adds a test for this case.

I have also squashed the older fixups.

leonardoalt commented on solc/CommandLineInterface.cpp:
```diff
@@ -451,7 +451,7 @@ bool CommandLineInterface::readInputFiles()
 			m_standardJsonInput = readUntilEnd(m_sin);
```
Not related but this line seems to have its indentation off. Could you fix it since you're touching the file already?

cameel replied:
Sure, but the indent on this line and others around it look fine to me (it's all tabs). Is this the right line?

leonardoalt replied:
Hm it looked weird on GitHub, not sure then.

cameel replied:
Yeah, github renders tabs in a weird way - as if they were shifted left relative to the beginning of the line. Really annoying.

leonardoalt commented on solc/CommandLineInterface.cpp:
```diff
@@ -451,7 +451,7 @@ bool CommandLineInterface::readInputFiles()
 		}
 		else
-			m_fileReader.setSource(g_stdinFileName, readUntilEnd(m_sin));
+			m_fileReader.setStdin(readUntilEnd(m_sin));
```
Is this the only place we read from stdin?

cameel replied:
Yeah. This and the one 3 lines up. It used to be done separately for Standard JSON but I unified that in #11544 so that it's now done here for all input modes.

leonardoalt commented on libsolidity/interface/FileReader.cpp:
```diff
@@ -31,9 +33,22 @@ using std::string;
+
+void FileReader::setStdin(SourceCode _source)
+{
+	m_sourceCodes["<stdin>"] = std::move(_source);
```
Where do we use `m_sourceCodes["<stdin>"]`? I'm wondering if the key `"<stdin>"` can get misused/misplaced somewhere

cameel replied:
It's passed along with all the other input files to the compiler stack. You can technically import it just like a normal file with `import "<stdin>"` because that's its source unit name. This is a pre-existing behavior documented in https://docs.soliditylang.org/en/latest/path-resolution.html#index-3.

It's possible to have a file that gets the same source unit name:
```bash
echo 'contract A {}' > '<stdin>'
echo 'contract B {}' | solc '<stdin>' - --base-path .
```
this line would then overwrite the content of that file in `FileReader` and replace it with the content of standard input. Not great but I think that `<stdin>` as a file name is obscure enough that it does not really matter all that much.

In any case this is how the compiler already worked before and my PR does not change that. I only added a dedicated function for adding content for this source unit name because `setSource()` now performs normalization and we obviously do not want that here.

leonardoalt replied:
Ok sounds good.

leonardoalt commented on test/solc/CommandLineInterface.cpp:
```diff
@@ -141,14 +141,19 @@ BOOST_AUTO_TEST_CASE(cli_input)
 		{"a", "b", "c/d/e/"},
 	};
 	map<string, string> expectedSources = {
 		{"<stdin>", "\n"},
```
Ah, here. Nevermind ;)

leonardoalt replied:
Ah no, this is a test. So my initial question still holds

cameel replied:
Answered in https://github.com/ethereum/solidity/pull/11545#pullrequestreview-736917746 above.

leonardoalt commented on test/solc/Common.cpp:
```diff
@@ -56,3 +56,13 @@ test::OptionsReaderAndMessages test::parseCommandLineAndReadInputFiles(
+	CommandLineOptions options;
+
+	options.modelChecker.initialize = true;
+	options.modelChecker.settings = {};
```
Why is `options.modelChecker` the only one that needs explicit assignment?

cameel replied:
It's a leftover from the CLI refactor. I was not sure if just letting it be initialized with defaults was equivalent to not initializing it at all so I added the `initialize` flag to exactly preserve the old behavior and straighten it out later, in a smaller PR. I later noticed that `initialize` always gets set anyway because model checker options have defaults so they are always present in the input.

So it's really only the `initialize` flag (I'll remove `settings = {}` as now it's clear it's not needed). The flag should be refactored out but I'm holding back with refactors until my big pile of PRs goes through. They almost always conflict. It's going to be eventually cleaned up as a part of #11727 anyway.

cameel replied:
Since the function now only sets this one flag and the flag is going to be removed eventually, I decided to just remove `defaultCommandLineOptions()`. The flag is now set directly in tests.

aarlt commented on libsolidity/interface/FileReader.h:
```diff
@@ -83,7 +88,43 @@ class FileReader
+	/// If @a _path and @_prefix are identical, the result is '.'.
+	static boost::filesystem::path stripPrefixIfPresent(boost::filesystem::path const& _prefix, boost::filesystem::path const& _path);
+
+	// @returns true if the specified path is an UNC path.
```
Missing doxygen triple slashes

aarlt commented on test/CMakeLists.txt:
```diff
@@ -103,6 +103,7 @@ set(libsolidity_sources
     libsolidity/SyntaxTest.h
     libsolidity/ViewPureChecker.cpp
     libsolidity/analysis/FunctionCallGraph.cpp
+    libsolidity/interface/FileReader.cpp
```
What about `libsolidity/interface/FileReader.h`?

cameel replied:
Boost tests typically do not have header files. Boost can find and run them without it.

The ones that do have headers here are actually a part of soltest/isoltest implementation and not test cases.

aarlt commented on test/FilesystemUtils.cpp:
```diff
@@ -54,11 +54,16 @@ void solidity::test::createFileWithContent(boost::filesystem::path const& _path,
 	boost::filesystem::path const& _targetPath,
-	boost::filesystem::path const& _linkName
+	boost::filesystem::path const& _linkName,
+	bool directorySymlink
```
`_directorySymlink`

aarlt commented on test/FilesystemUtils.h:
```diff
@@ -39,12 +39,16 @@ void createFileWithContent(boost::filesystem::path const& _path, std::string con
 	boost::filesystem::path const& _targetPath,
-	boost::filesystem::path const& _linkName
+	boost::filesystem::path const& _linkName,
+	bool directorySymlink
```
`_directorySymlink`
